### DISCUSSION TITLE
MIDI routing

### DISCUSSION THREAD

**Discussion by splisp** [falkTX/Carla#1561] [Q&A]
I am using the Windows version. As I can understand, in "Patchbay" mode there are only a single MIDI input and a single MIDI output available. I can't understand which ones, out of the available system I/O. There is no "MIDI setup" where I can define the MIDI devices that Carla is going to open. Does she open all of the devices? What if I want to use some devices for other programs? Windows doesn't support multi-client MIDI devices, AFAIK.

So, in order to define a custom MIDI routing, I set the engine to "Continous Rack" mode. Then I am presented with a patchbay (and this is confusing) where I find a "Carla" node. Double-clicking it has no effect. I can add other "Carla patchbay" items in the rack but they don't appear in the patchbay view; I can open the window of the first Carla-patchbay item in the rack by clicking on the gear icon, but the second one only opens if I disable the first one. *EDIT* My bad, looks like I can open the windows for all the Carla-patchbay items in the rack

I can see all the MIDI I/O in the patchbay when in Rack mode and route single MIDI inputs to... the Carla node's input. Can I add MIDI filters inbetween the inputs and the Carla node? Is the routing (direct or filtered) applied to the first Carla-patchbay in the rack, then output of the first patchbay (if any) is sent to the next plugin in the rack (possibly a second Carla-patchbay)?

**Comment by splisp** [comment]
Regarding the problem of being unable to open the window for some instance of Carla-patchbay in the rack: sometimes, quite randomly, the window fails to show and the stdout outputs

waitForClientFirstMessage() - read failed

Then the gear icon gets disabled, and in order to re-enable it I have to collapse/expand the rack a couple of times.

**Comment by splisp** [comment]
I made up my idea that the "Carla" node in the patchbay, when the engine is in "Continuous Rack" mode, is actually the whole rack. In the patchbay, I can configure which audio and MIDI inputs are connected to the first device in the rack, and which audio and MIDI outputs are connected to the last device in the rack. This is not so much intuitive, moreso since the patchbay view is used in a different way when in "Patchbay" mode, but once I know it it's fine!

**Comment by falkTX** [comment]
Under patchbay mode you need to use the menu canvas -> show external and connect things from there.
The MIDI setup is kinda hidden behind that step, it was a way I found to support such MIDI setup without needing to implement specific dialogs for it.

**Reply by splisp** [reply]
I see now. It's a bit weird but once one knows it it makes sense. Just a small glitch: the External Canvas view doesn't retain its design (positions of the nodes are reset after switching to Internal view and back to External).
